Build trouble in VSCode

Hi! I'm trying to follow the logic in the GitHub workflow and I have this Dockerfile:

```dockerfile
FROM ubuntu:20.04

LABEL maintainer="Vanessasaurus <@vsoch>"

# Match the default user id for a single system so we aren't root
ARG USERNAME=vscode
ARG USER_UID=1000
ARG USER_GID=1000
ENV USERNAME=${USERNAME}
ENV USER_UID=${USER_UID}
ENV USER_GID=${USER_GID}
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt install -y clang llvm-dev libjansson-dev libssl-dev \
    mpi-default-dev wget bison flex make cmake mpich python3 python3-pip sudo llvm-12 
    # Note I didn't get it working with just mpich!
#    openmpi-bin openmpi-common libopenmpi-dev libgtk2.0-dev

# install cuda 12.1 (note uncomment this if you can support on your development machine!)
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin && \
    mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
    wget https://developer.download.nvidia.com/compute/cuda/12.1.0/local_installers/cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
    dpkg -i cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
    cp /var/cuda-repo-ubuntu2004-12-1-local/cuda-*-keyring.gpg /usr/share/keyrings/ && \
    apt-get update && \
    apt-get -y install cuda && \
    ln -s -T /usr/bin/make /usr/bin/gmake

ENV PATH=/usr/local/cuda-12.1/bin:$PATH

# Python helpers
RUN python3 -m pip install --upgrade pip pytest setuptools flake8 rstfmt black
#    sudo update-alternatives --install /usr/bin/llvm-config llvm-config /usr/bin/llvm-config-12 200

RUN ldconfig

# Add the group and user that match our ids
RUN groupadd -g ${USER_GID} ${USERNAME} && \
    adduser --disabled-password --uid ${USER_UID} --gid ${USER_GID} --gecos "" ${USERNAME} && \
    echo "${USERNAME} ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers
USER $USERNAME
```

And then to build:

```console
export PYTHONPATH=$PWD/src/python:$PYTHONPATH
cd src/c
mkdir -p build install
cd build
export CMAKE_OPTS="-DCMAKE_CXX_COMPILER=clang++ -DLLVM_DIR=/usr/lib/llvm-12/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../install"
cmake ${CMAKE_OPTS} ..
```

But the build fails - here is the issue

```console
Determining if the C compiler works failed with the following output:
Change Dir: /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_599b9/fast && No such file or directory
Generator: execution of make failed. Make command was: /usr/bin/gmake cmTC_599b9/fast && 


The MPI test test_mpi for CXX in mode normal failed to compile with the following output:
Change Dir: /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_21bda/fast && /usr/bin/gmake -f CMakeFiles/cmTC_21bda.dir/build.make CMakeFiles/cmTC_21bda.dir/build
gmake[1]: Entering directory '/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o
/usr/bin/clang++   -isystem /usr/include/x86_64-linux-gnu/mpich  -fPIE   -flto=auto -ffat-lto-objects -o CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o -c /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/FindMPI/test_mpi.cpp
clang: error: unsupported argument 'auto' to option 'flto='
clang: warning: optimization flag '-ffat-lto-objects' is not supported [-Wignored-optimization-argument]
gmake[1]: *** [CMakeFiles/cmTC_21bda.dir/build.make:66: CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o] Error 1
gmake[1]: Leaving directory '/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:121: cmTC_21bda/fast] Error 2
```

Is there a flag or something I;m missing on a path or a version error? I'm using ubuntu 20.04 and mpich like the workflow file. Thanks!

This is really strange - when I build the exact same container from VSCode, I get a different error:

```
-- Setting build type to "Debug"
-- CMAKE_INSTALL_RPATH = /workspaces/PerfFlowAspect/src/c/install/lib
-- Could NOT find MPI_CXX (missing: MPI_CXX_WORKS) 
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find MPI (missing: MPI_CXX_FOUND) (found version "3.1")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1688 (find_package_handle_standard_args)
  test/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeOutput.log".
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeError.log".
```

I tried as root (as I did above) and the vscode user (same issue!) I think this was an original error I saw, so maybe we should debug from here. So it's telling me MPI isn't installed - but I definitely installed mpich. Here are the paths to what I found in /usr with mpi:

```
$ find . -name *mpi*.so
./lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so
./lib/x86_64-linux-gnu/libmpich.so
./lib/x86_64-linux-gnu/libompitrace.so
./lib/x86_64-linux-gnu/libmpichfort.so
./lib/x86_64-linux-gnu/libmpichcxx.so
./lib/x86_64-linux-gnu/openmpi/lib/libompitrace.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/libompi_dbg_msgq.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_schizo_ompi.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_io_ompio.so
./lib/x86_64-linux-gnu/openmpi/lib/libmca_common_ompio.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_java.so
./lib/x86_64-linux-gnu/openmpi/lib/ompi_monitoring_prof.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempi_ignore_tkr.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_mpifh.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempif08.so
./lib/x86_64-linux-gnu/libmca_common_ompio.so
./lib/x86_64-linux-gnu/libmpi_java.so
./lib/x86_64-linux-gnu/libmpi_usempi_ignore_tkr.so
./lib/x86_64-linux-gnu/open-coarrays/openmpi/lib/libcaf_mpi.so
./lib/x86_64-linux-gnu/open-coarrays/openmpi/lib/libcaf_openmpi.so
./lib/x86_64-linux-gnu/libmpi_mpifh.so
./lib/x86_64-linux-gnu/libmpi_cxx.so
./lib/x86_64-linux-gnu/libmpi.so
./lib/x86_64-linux-gnu/libcaf_openmpi.so
./lib/x86_64-linux-gnu/libmpi_usempif08.so
./lib/x86_64-linux-gnu/libmpi++.so
```
What should I try?

okay this gives us some insights! It seems it _was_ targeting openmpi, because now it's angry that it's missing! :laughing: 

```
CMake Error in /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
  Imported target "MPI::MPI_C" includes non-existent path

    "/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error at /usr/share/cmake-3.16/Modules/FindMPI.cmake:1194 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1245 (_MPI_try_staged_settings)
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1505 (_MPI_check_lang_works)
  test/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeOutput.log".
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeError.log".
```

--- a/.devcontainer/Dockerfile
+++ b/.devcontainer/Dockerfile
@@ -1,0 +1,47 @@
+FROM ubuntu:20.04
+
+# export PATH=/usr/local/cuda-12.1/bin/:$PATH
+# cd src/c
+# mkdir build
+# cd build
+# cmake -DCMAKE_CXX_COMPILER=clang++ -DLLVM_DIR=/usr/lib/llvm-10/cmake ..
+
+LABEL maintainer="Vanessasaurus <@vsoch>"
+
+# Match the default user id for a single system so we aren't root
+ARG USERNAME=vscode
+ARG USER_UID=1000
+ARG USER_GID=1000
+ENV USERNAME=${USERNAME}
+ENV USER_UID=${USER_UID}
+ENV USER_GID=${USER_GID}
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update && \
+    apt install -y clang llvm-dev libjansson-dev libssl-dev \
+    wget bison flex make cmake mpich python3 python3-pip sudo llvm-12 \
+    sudo pciutils
+
+# install cuda 12.1 (note uncomment this if you can support on your development machine!)
+RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin && \
+    mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
+    wget https://developer.download.nvidia.com/compute/cuda/12.1.0/local_installers/cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
+    dpkg -i cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
+    cp /var/cuda-repo-ubuntu2004-12-1-local/cuda-*-keyring.gpg /usr/share/keyrings/ && \
+    apt-get update && \
+    apt-get -y install cuda && \
+    ln -s -T /usr/bin/make /usr/bin/gmake
+
+ENV PATH=/usr/local/cuda-12.1/bin:$PATH
+
+# Python helpers
+RUN python3 -m pip install --upgrade pip pytest setuptools flake8 rstfmt black
+#    sudo update-alternatives --install /usr/bin/llvm-config llvm-config /usr/bin/llvm-config-12 200
+
+RUN ldconfig
+
+# Add the group and user that match our ids
+RUN groupadd -g ${USER_GID} ${USERNAME} && \
+    adduser --disabled-password --uid ${USER_UID} --gid ${USER_GID} --gecos "" ${USERNAME} && \
+    echo "${USERNAME} ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers
+USER $USERNAME

--- a/.devcontainer/devcontainer.json
+++ b/.devcontainer/devcontainer.json
@@ -1,0 +1,5 @@
+{
+    "name": "PerfFlowAspect Development Environment",
+    "dockerFile": "Dockerfile",
+    "context": "../"
+}

--- a/.github/workflows/build-deploy.yaml
+++ b/.github/workflows/build-deploy.yaml
@@ -1,0 +1,72 @@
+name: build PerfFlowAspect
+
+on:
+
+  # Publish packages on release
+  release:
+    types: [published]
+
+  # We might want to remove this, OR provide a base container
+  # that will more quickly build from, or use cache
+  pull_request: []
+
+  # On push to main we build and deploy images
+  push:
+    branches:
+    - main
+
+jobs:
+  build:
+    permissions:
+      packages: write
+
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        container: [[".devcontainer/Dockerfile", ".", "ghcr.io/flux-framework/perf-flow-aspect-base"]]
+        # TODO add the container with PerfFlowAspect after base is built
+
+    name: Build ${{ matrix.container[2] }}
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+      
+    - name: Make Space For Build
+      run: |
+        sudo rm -rf /usr/share/dotnet
+        sudo rm -rf /usr/local/lib/android
+        sudo rm -rf /opt/ghc
+
+    - name: Pull Layers
+      env:
+        container: ${{ matrix.container[2] }}
+      run: docker pull ${container} || echo "No layers to pull for ${container}"
+
+    - name: Build Container
+      env:
+        dockerfile: ${{ matrix.container[0] }}
+        container: ${{ matrix.container[2] }}
+        context: ${{ matrix.container[1] }}
+      run: docker build -f ${dockerfile} -t ${container} ${context}
+
+    - name: GHCR Login
+      if: (github.event_name != 'pull_request')
+      uses: docker/login-action@v2
+      with:
+        registry: ghcr.io
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+
+    - name: Tag and Push Release Image
+      if: (github.event_name == 'release')
+      env:
+        container: ${{ matrix.container[2] }}
+      run: |
+        tag=${GITHUB_REF#refs/tags/}
+        echo "Tagging and releasing ${container}:${tag}"
+        docker tag ${container}:latest ${container}:${tag}
+        docker push ${container}:${tag}
+
+    - name: Deploy
+      if: (github.event_name != 'pull_request')
+      run: docker push ${container}:latest

--- a/.gitignore
+++ b/.gitignore
@@ -1,3 +1,4 @@
+src/c/build
 src/c/parser/lex.yy.c
 src/c/parser/lex.yy.o
 src/c/parser/libperfflow_parser.so

--- a/docs/developer.rst
+++ b/docs/developer.rst
@@ -1,0 +1,73 @@
+############################
+ PerfFlowAspect Development
+############################
+
+This is a short developer guide for using the included development environment.
+We provide both a base container for `VSCode
+<https://code.visualstudio.com/docs/remote/containers>`_, and a production
+container with PerfFlowAspect you can use for application development outside of
+that. This is done via the `.devcontainer
+<https://code.visualstudio.com/docs/remote/containers#_create-a-devcontainerjson-file`_
+directory. You can follow the `DevContainers tutorial
+<https://code.visualstudio.com/docs/remote/containers-tutorial>`_ where you'll
+basically need to:
+
+#. Install Docker, or compatible engine
+#. Install the Development Containers extension
+
+Then you can go to the command palette (View -> Command Palette) and select
+``Dev Containers: Open Workspace in Container.`` and select your cloned
+PerfFlowAspect repository root. This will build a development environment. You
+are free to change the base image and rebuild if you need to test on another
+operating system! When your container is built, when you open ``Terminal -> New
+Terminal`` you'll be in the container, which you can tell based on being the
+"vscode" user. You can then proceed to the sections below to build and test
+PerfFlowAspect.
+
+**Important** the development container assumes you are on a system with uid
+1000 and gid 1000. If this isn't the case, edit the ``.devcontainer/Dockerfile``
+to be your user and group id. This will ensure changes written inside the
+container are owned by your user. It's recommended that you commit on your
+system (not inside the container) because if you need to sign your commits, the
+container doesn't have access and won't be able to. If you find that you
+accidentally muck up permissions and need to fix, you can run this from your
+terminal outside of VSCode:
+
+.. code:: console
+
+   $ sudo chown -R $USER .git/
+   # and then commit
+
+***************************
+ Installing PerfFlowAspect
+***************************
+
+Once inside the development environment, you can compile PerfFlowAspect:
+
+.. code:: console
+
+   export PATH=/usr/local/cuda-12.1/bin/:$PATH
+   cd src/c
+   mkdir build
+   cd build
+   cmake -DCMAKE_CXX_COMPILER=clang++ -DLLVM_DIR=/usr/lib/llvm-10/cmake ..
+   make
+   sudo make install
+
+If you want to run tests, cd to where the tests are, and run a few!
+
+.. code:: console
+
+   cd src/c/build/test
+   ./t0001-cbinding-basic.t
+
+Note that if you don't have a GPU, these probably will error (I do not)! Here is
+how to run Python tests:
+
+.. code:: console
+
+   cd src/python/test
+   ./t0001-pybinding-basic.t
+
+You'll again have issues without an actual GPU. And that's it! Note that we will
+update this documentation as we create more examples.

